test/hotspot/gtest/nmt/test_vmatree.cpp

```diff
@@ -284,6 +284,68 @@ TEST_VM_F(NMTVMATreeTest, LowLevel) {
   }
 }
 
+TEST_VM_F(NMTVMATreeTest, SetTag) {
+  auto i = [](MemTag f) -> uint8_t { return (uint8_t)f; };
+
+  // The gc/cds case with only reserved data
+  {
+    VMATree::SummaryDiff diff;
+    Tree::RegionData rd(NCS::StackIndex(), mtNone);
+    VMATree tree;
+
+    VMATree::SummaryDiff result = tree.reserve_mapping(0, 500, rd);
+    diff.add(result);
+    EXPECT_EQ(500, diff.tag[i(mtNone)].reserve);
+
+    result = tree.reserve_mapping(500, 100, rd);
+    diff.add(result);
+    EXPECT_EQ(600, diff.tag[i(mtNone)].reserve);
+
+    result = tree.set_tag(0, 500, mtGC);
+    diff.add(result);
+    EXPECT_EQ(100, diff.tag[i(mtNone)].reserve);
+    EXPECT_EQ(500, diff.tag[i(mtGC)].reserve);
+
+    result = tree.set_tag(500, 100, mtClassShared);
+    diff.add(result);
+    EXPECT_EQ(0, diff.tag[i(mtNone)].reserve);
+    EXPECT_EQ(500, diff.tag[i(mtGC)].reserve);
+    EXPECT_EQ(100, diff.tag[i(mtClassShared)].reserve);
```

Comment:
As  it is now, the test only compares sums, and needs quite a bit of boilerplate coding for that.

My proposal would be:

- give VMATree a test function that takes an array of ranges (plain simple stupid POD structs) and compares them item by item with its internal state. Let it return true on success, print an error on an outputStream* we give the function on error.
- then, in the test, hardcode the test array and let VMATree compare it.

Example call for above test:
```
struct testrange { from, to, stackindex, tag };
const testrange[] expected_form = {
 { 0, 500, NCS::StackIndex(), mtGC },
 { 500, 600, NCS::StackIndex(), mtClassShared },
};
tree.test_against(expected_form, tty);
```

That would be easier on the eye in the test file, more succinct, and catch more errors. Because we compare actual addresses, and we also compare expected stack indices. As it is now, this test would not catch if we accidentally screw up stacks.